tests/test_chi_squared_versus_gsl.cc

```diff
@@ -0,0 +1,5742 @@
+/*
+ * Copyright (C) 2024 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <albatross/Common>
+#include <albatross/Stats>
+
+#include <gtest/gtest.h>
+
+namespace albatross {
+
+/*
+  What follows is a long list of spot checks of our incomplete gamma
+  implementation. The function is critical for the evaluation of the chi squared
+  percentiles and since those values often factor in to outlier detection we
+  need to be sure we have good precision on the tails. What follows were a bunch
+  of comparisons to the GSL implementations which were generated using:
+
+  #include <gsl/gsl_sf_gamma.h>
+
+  double dof = 1.0;
+  while (dof < 2000) {
+    const double a = 0.5 * dof;
+
+    const auto bounds =
+        details::incomplete_gamma_quadrature_bounds(a, 100 * sqrt(a));
+    const double lowest = 2. * bounds.first;
+    const double highest = 2. * bounds.second;
+
+    auto gsl_chi_squared = [](double x, double dof) {
+      return gsl_sf_gamma_inc_P(0.5 * dof, 0.5 * x);
+    };
+
+    // The bounds provided should correspond to the z argument which corresponds
+    // to near zero and near one evaluations of the incomplete gamma
+    EXPECT_NEAR(chi_squared_cdf_non_integer_dof(lowest, dof), 0., 1e-8);
+    EXPECT_NEAR(chi_squared_cdf_non_integer_dof(highest, dof), 1., 1e-8);
+
+    for (const auto &x : linspace(lowest, highest, 101)) {
+      const double actual = chi_squared_cdf_non_integer_dof(x, dof);
+      const double expected = gsl_chi_squared(x, dof);
+      std::cout << std::setprecision(12);
+      // std::cout << "UNITTEST: EXPECT_NEAR(chi_squared_cdf_non_integer_dof("
+  << x << ", " << dof
+      //           << "), " << expected << ", 1e-6);" << std::endl;
+      std::cout << "UNITTEST: {" << x << ", " << dof << ", " << expected << "},"
+  << std::endl;
+    }
+
+    // explore the lower degrees of freedom better
+    if (dof > 50) {
+      dof *= 2;
+    } else {
+      dof += 1;
+    }
+  }
```

Comment:
Any way you can incorporate this function into the test to automatically dump out the values for use in the comparisons inside the `test_chi_squared_matches_gsl` unit tests? Might make any future modifications/iterations easier

Reply:
The problem with that is that we'd then need to pull GSL in as a third party library and their license (GNU public) is too restrictive for albatross. I suppose it's possible they'd allow use of the library as a test/comparison not as part of the core algorithms ... but that seems like a pretty gray area.

Reply:
Ugh never mind then.. we can stick it here https://github.com/swift-nav/shared-scratch in the future if we ever need to tweak it in the future

Reply:
Can you add a note to that comment so we remember why we couldn't just add that output function there? Or just copy and paste into https://github.com/swift-nav/shared-scratch and add a link from orion-engine there

Reply:
oops, sorry already merged it. but I'll follow up.

Reply:
https://github.com/swift-nav/albatross/compare/akleeman/incomplete_gamma_test_comment?expand=1